lib/src/license_detection/license_detector.dart

```diff
@@ -0,0 +1,244 @@
+
+    identifierToLicense[match.identifier] = match;
+  }
+
+  return identifierToLicense.values.toList();
+}
+
+/// Custom comparator to the sort the licenses based on decreasing order of confidence.
+///
+/// Incase the confidence detected is same for the matches, ratio of tokens claimed
+/// in the unkown text to the number of tokens present in the known license text
+/// is considered
+@visibleForTesting
+int sortOnConfidence(LicenseMatch matchA, LicenseMatch matchB) {
+  if (matchA.confidence > matchB.confidence) {
+    return -1;
+  }
+
+  if (matchA.confidence < matchB.confidence) {
+    return 1;
+  }
+
+  final matchATokensPercent =
+      matchA.tokensClaimed / matchA.license.tokens.length;
+  final matchBTokensPercent =
+      matchB.tokensClaimed / matchB.license.tokens.length;
+
+  return (matchATokensPercent > matchBTokensPercent) ? -1 : 1;
+}
+
+/// Fliters out licenses having overlapping ranges giving preferences to a match with higher token density.
+///
+/// Token density is the product of number of tokens claimed in the range and
+/// confidence score of the match. Incase of exact match we retain both
+/// the matches so that the user can resolve them.
+@visibleForTesting
+List<LicenseMatch> removeOverLappingMatches(List<LicenseMatch> matches) {
+  var retain = List.filled(matches.length, false);
+  var retainedmatches = <LicenseMatch>[];
+
+  // We consider token density to retain matches of larger licenses
+  // having lesser confidence when compared to a smaller license
+  // haing a  perfect match.
+  for (var i = 0; i < matches.length; i++) {
+    var keep = true;
+    final matchA = matches[i];
+    final rangeA = Range(matchA.start, matchA.end);
+
+    var proposals = <int, bool>{};
+    for (var j = 0; j < matches.length; j++) {
+      if (j == i) {
+        break;
+      }
+      final matchB = matches[j];
+      final rangeB = Range(matchB.start, matchB.end);
+      // Check if matchA is larger license containing an insatnce of
+      // smaller license within it and decide to whether retain it
+      // or not by comapring their token densities. Example NPL
+      // contains MPL.
+      if (rangeA.contains(rangeB) && retain[j]) {
+        final aConf = matchA.tokensClaimed * matchA.confidence;
+        final bConf = matchB.tokensClaimed * matchB.confidence;
+
+        // Retain both the licenses incase of a exact match,
+        // so that it can be resolved by the user.
+        if (aConf > bConf) {
+          proposals[j] = true;
+        } else if (bConf > aConf) {
+          keep = false;
+        }
+      } else if (rangeA.overlapsWith(rangeB)) {
+        keep = false;
+      }
+    }
+
+    if (keep) {
+      retain[i] = true;
+      proposals.forEach((key, value) {
+        retain[key] = value;
+      });
+    }
+  }
+
+  for (var i = 0; i < matches.length; i++) {
+    if (retain[i]) {
+      retainedmatches.add(matches[i]);
+    }
+  }
+
+  return retainedmatches;
+}
+
+double unclaimedTokenPercentage(
```

Comment:
This method is a bit confusing and without documentation: it takes matches with ranges, but it divides their sum with a separate count. Can it be more than 1.0? Please expand on it.
It also matches a field name, so maybe rename this to `calculateUnclaimedTokenPercentage`.